contracts/erc721/ERC721PreApproval.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2021 the ethier authors (github.com/divergencetech/ethier)
+pragma solidity >=0.8.0 <0.9.0;
+
+import "../thirdparty/opensea/OpenSeaGasFreeListing.sol";
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+
+/// @notice Pre-approval of OpenSea proxies for gas-less listing
+/// @dev This wrapper allows users to revoke the pre-approval of their
+/// associated proxy and emits the corresponding events. This is necessary for
+/// external tools to index approvals correctly and inform the user.
+/// @dev The pre-approval is triggered on a per-wallet basis during the first
+/// transfer transactions. It will only be enabled for wallets with an existing
+/// proxy. Not having a proxy incurs a gas overhead.
+/// @dev This wrapper optimizes for the following scenario:
+/// - The majority of users already have a wyvern proxy
+/// - Most of them want to transfer tokens via wyvern exchanges
+abstract contract ERC721PreApproval is ERC721 {
+    /// @dev It is important that Active remains at first position, since this
+    /// is the scenario that we are trying to optimize for.
+    enum State {
+        Active,
+        Inactive
+    }
+
+    /// @notice The state of the pre-approval for a given address
+    mapping(address => State) private state;
+
+    /// @dev Returns true if either standard `isApprovedForAll()` or if the
+    /// `operator` is the OpenSea proxy for the `owner` provided the
+    /// pre-approval is active.
+    function isApprovedForAll(address owner, address operator)
+        public
+        view
+        virtual
+        override
+        returns (bool)
+    {
+        if (super.isApprovedForAll(owner, operator)) {
+            return true;
+        }
+
+        if (state[owner] == State.Active) {
+            return OpenSeaGasFreeListing.isApprovedForAll(owner, operator);
+        }
+
+        return false;
```

Comment:
Why not just `return state[owner] == State.Active && OpenSeaGas…`?

Reply:
Afaik `OpenSeaGasFreeListing.isApprovedForAll` is called regardless of the outcome of `state[owner] == State.Active` in an `.. && ..` statement. So the idea was to save some gas for `Inactive`.

However, this might introduce a small overhead for `Active` (would have to measure it), which goes against the initial assumptions and optimization goals. Therefore, I will change it to a `.. && ..` pattern for now.

Reply:
Solidity appears to have [short-circuit boolean logic](https://github.com/ethereum/solidity/issues/576).

Reply:
Oh cool, I was not aware of that! Thanks